frontend/catalyst/compilation_pipelines.py

```diff
@@ -214,36 +251,22 @@ def _exec(shared_object_file, func_name, has_return, numpy_dict, *args):
             retval: the value computed by the function or None if the function has no return value
         """
 
-        shared_object, function, setup, teardown, mem_transfer = CompiledFunction.load_symbols(
-            shared_object_file, func_name
-        )
-
         params_to_setup = [b"jitted-function"]
         argc = len(params_to_setup)
         array_of_char_ptrs = (ctypes.c_char_p * len(params_to_setup))()
         array_of_char_ptrs[:] = params_to_setup
 
-        setup(ctypes.c_int(argc), array_of_char_ptrs)
+        lib.setup(ctypes.c_int(argc), array_of_char_ptrs)
         result_desc = type(args[0].contents) if has_return else None
 
-        retval = wrapper.wrap(function, args, result_desc, mem_transfer, numpy_dict)
+        retval = wrapper.wrap(lib.function, args, result_desc, lib.mem_transfer, numpy_dict)
         if len(retval) == 0:
             retval = None
         elif len(retval) == 1:
             retval = retval[0]
 
         # Teardown has to be made after the return valued has been copied.
-        teardown()
-
-        # Unmap the shared library. This is necessary in case the function is re-compiled.
-        # Without unmapping the shared library, there would be a conflict in the name of
-        # the function and the previous one would succeed.
-        # Need to close after obtaining value, since the value can point to memory in the shared
-        # object. This is in the case of returning a constant, for example.
-        dlclose = ctypes.CDLL(None).dlclose
-        dlclose.argtypes = [ctypes.c_void_p]
-        # pylint: disable=protected-access
-        dlclose(shared_object._handle)
+        lib.teardown()
```

Comment:
This place catches my eye: Should we use `try-finally` here to behave well in case of exceptions?

Reply:
What exception are you thinking we should catch here? And what behaviour would be good in case we cannot close the shared library? I think this is technically an unrecoverable error at the moment, the alternative would be to create a shared object for every possible compiled function and just keep loading them.

Reply:
Yes, I am talking about the `teardown`, but not about catching exceptions. Assuming that `lib.setup` does some kind of resource allocation we might want to free, I expected to see something like this:
``` python
lib.setup()
try:
  somehow_call(lib.function)
...
finally:
  lib.teardown()
```

This way I hope to get sure that Python would free the resource even if lib.function or its wrapper raise some exception.

Reply:
Awesome suggestion: see here https://github.com/PennyLaneAI/catalyst/pull/213/commits/c42d1317da8b05908611d5391f091df3836a533a